src/hotspot/cpu/aarch64/macroAssembler_aarch64.hpp

```diff
@@ -437,11 +437,39 @@ class MacroAssembler: public Assembler {
     Assembler::INSN(Rd, Rn, Rm, Ra);                                          \
   }
 
-  WRAP(madd) WRAP(msub) WRAP(maddw) WRAP(msubw)
+  WRAP(madd) WRAP(maddw)
   WRAP(smaddl) WRAP(smsubl) WRAP(umaddl) WRAP(umsubl)
 #undef WRAP
 
 
+  inline void msub(Register Rd, Register Rn, Register Rm, Register Ra) {
+    if (VM_Version::supports_a53mac() && Ra != zr)
+      nop();
+    if (VM_Version::is_neoverse_n_series()) {
+      /* On Neoverse N series, MSUB uses the same ALU with SDIV.
+       * The combination of MUL/SUB can utilize multiple ALUS,
+       * and is much faster than MSUB. */
+      mul(rscratch1, Rn, Rm);
+      sub(Rd, Ra, rscratch1);
+    } else {
+      Assembler::msub(Rd, Rn, Rm, Ra);
+    }
+  }
+
+  inline void msubw(Register Rd, Register Rn, Register Rm, Register Ra) {
+    if (VM_Version::supports_a53mac() && Ra != zr)
+      nop();
+    if (VM_Version::is_neoverse_n_series()) {
```

Comment:
Why only Neoverse N series? Even on the V series (V1 and V2), both `sdiv/udiv` and `msub` instructions are executed in M0 unit (Integer multi cycle). It should benefit the V series as well. Source: https://developer.arm.com/documentation/pjdoc466751330-9685/latest/ and https://developer.arm.com/documentation/PJDOC-466751330-593177/latest/

A quick run on a V1 machine shows ~15% performance gain for the `IntegerDivMod` tests if we generate separate `mul` and `sub` instructions instead of a single `msub`.

Reply:
Thanks for your review.
This new commit includes the support for V1/V2 you mentioned.

https://github.com/openjdk/jdk/pull/19093/commits/d8b8dbfe102d2716ef9e332aec7c52e566bf1727

Reply:
> Why only Neoverse N series? Even on the V series (V1 and V2), both `sdiv/udiv` and `msub` instructions are executed in M0 unit (Integer multi cycle). It should benefit the V series as well. Source: https://developer.arm.com/documentation/pjdoc466751330-9685/latest/ and https://developer.arm.com/documentation/PJDOC-466751330-593177/latest/
> 
> A quick run on a V1 machine shows ~15% performance gain for the `IntegerDivMod` tests if we generate separate `mul` and `sub` instructions instead of a single `msub`.

Not sure if this can benefit V3, since MSUB can use M rather M0.

Reply:
And it might also be that separate mul/sub neither helps nor hurts V3. Let's not over-engineer this patch.